test/com/linkedin/drelephant/tuning/TuningManagerTest.java

```diff
@@ -47,49 +47,60 @@ public void onStart(Application app) {
     fakeApp = fakeApplication(dbConn, gs);
     Configuration configuration = ElephantContext.instance().getAutoTuningConf();
     Boolean autoTuningEnabled = configuration.getBoolean(DrElephant.AUTO_TUNING_ENABLED, false);
-   // org.junit.Assume.assumeTrue(autoTuningEnabled);
+    // org.junit.Assume.assumeTrue(autoTuningEnabled);
   }
 
   @Test
   public void testIPSOManager() {
     running(testServer(TEST_SERVER_PORT, fakeApp), new IPSOManagerTestRunner());
-
   }
 
   @Test
-  public void testFlowTestRunner(){
+  public void testFlowTestRunner() {
     running(testServer(TEST_SERVER_PORT, fakeApp), new FlowTestRunner());
   }
 
-
   @Test
-  public void testBaselineManagerTestRunner(){
+  public void testBaselineManagerTestRunner() {
     running(testServer(TEST_SERVER_PORT, fakeApp), new BaselineManagerTestRunner());
   }
 
   @Test
-  public void testJobStatusManagerTestRunner(){
+  public void testJobStatusManagerTestRunner() {
     running(testServer(TEST_SERVER_PORT, fakeApp), new JobStatusManagerTestRunner());
   }
 
-
   @Test
-  public void testFitnessManagerTestRunner(){
+  public void testFitnessManagerTestRunner() {
     running(testServer(TEST_SERVER_PORT, fakeApp), new FitnessManagerTestRunner());
   }
 
   @Test
-  public void testParamGenerterTestRunner(){
+  public void testParamGenerterTestRunner() {
     running(testServer(TEST_SERVER_PORT, fakeApp), new ParameterGenerateManagerTestRunner());
   }
 
   @Test
-  public void testAlgoBasedOnVersion(){
-    assertTrue("Alorithm Based on Version Test", controllers.Application.getAlgoBasedOnVersion(1).equals(
-        TuningAlgorithm.OptimizationAlgo.PSO_IPSO.name()));
-    assertTrue("Alorithm Based on Version Test", controllers.Application.getAlgoBasedOnVersion(2).equals(
-        TuningAlgorithm.OptimizationAlgo.HBT.name()));
-    assertTrue("Alorithm Based on Version Test", controllers.Application.getAlgoBasedOnVersion(3).equals(
-        TuningAlgorithm.OptimizationAlgo.HBT.name()));
+  public void testAlgoBasedOnVersion() {
+    assertTrue("Alorithm Based on Version Test",
+        controllers.Application.getAlgoBasedOnVersion(1).equals(TuningAlgorithm.OptimizationAlgo.PSO_IPSO.name()));
+    assertTrue("Alorithm Based on Version Test",
+        controllers.Application.getAlgoBasedOnVersion(2).equals(TuningAlgorithm.OptimizationAlgo.HBT.name()));
+    assertTrue("Alorithm Based on Version Test",
+        controllers.Application.getAlgoBasedOnVersion(3).equals(TuningAlgorithm.OptimizationAlgo.HBT.name()));
+  }
+
+  @Test
+  public void testAutoTunerApiHelper() throws InterruptedException {
+    for (int i = 0; i < 50; i++) {
+      TuningInput tuningInput = new TuningInput();
+      tuningInput.setFlowExecId(1 + "");
+      tuningInput.setFlowExecUrl(1 + "");
+      tuningInput.setFlowDefId(i + "");
+      tuningInput.setFlowDefUrl(i + "");
+      Thread t1 = new Thread(new AutoTunerApiTestRunner(new AutoTuningAPIHelper(), tuningInput));
```

Comment:
This code doesn't really test the race condition. Here threads will execute in a sequence. That is thread 1 will start and complete, then only thread 2 will start and complete, and so on. join on t1 will ensure it waits for the thread to complete.
We can instead use ExecutorService, call shutdown and awaitTermination on it with a suitable timeout to test this.

Reply:
I have tested the race condition by putting sleep in getFlowExecution and started two threads together and then joined them (but removed sleep from actual code) . I'll change it to ExecutorService to check  for race condition